src/plugins/dashboard/public/application/actions/clone_panel_action.tsx

```diff
@@ -0,0 +1,166 @@
+import uuid from 'uuid';
+import { ActionByType, IncompatibleActionError } from '../../ui_actions_plugin';
+import { ViewMode, PanelState, IEmbeddable } from '../../embeddable_plugin';
+import { SavedObject } from '../../../../saved_objects/public';
+import { PanelNotFoundError, EmbeddableInput } from '../../../../embeddable/public';
+import {
+  placePanelBeside,
+  IPanelPlacementBesideArgs,
+} from '../embeddable/panel/dashboard_panel_placement';
+import { DashboardPanelState, DASHBOARD_CONTAINER_TYPE, DashboardContainer } from '..';
+
+export const ACTION_CLONE_PANEL = 'clonePanel';
+
+export interface ClonePanelActionContext {
+  embeddable: IEmbeddable;
+}
+
+export class ClonePanelAction implements ActionByType<typeof ACTION_CLONE_PANEL> {
+  public readonly type = ACTION_CLONE_PANEL;
+  public readonly id = ACTION_CLONE_PANEL;
+  public order = 11;
+
+  constructor(private core: CoreStart) {}
+
+  public getDisplayName({ embeddable }: ClonePanelActionContext) {
+    if (!embeddable.getRoot() || !embeddable.getRoot().isContainer) {
+      throw new IncompatibleActionError();
+    }
+    return i18n.translate('dashboard.panel.clonePanel', {
+      defaultMessage: 'Clone panel',
+    });
+  }
+
+  public getIconType({ embeddable }: ClonePanelActionContext) {
+    if (!embeddable.getRoot() || !embeddable.getRoot().isContainer) {
+      throw new IncompatibleActionError();
+    }
+    return 'copy';
+  }
+
+  public async isCompatible({ embeddable }: ClonePanelActionContext) {
+    if (embeddable.getInput().viewMode) {
+      if (embeddable.getInput().viewMode === ViewMode.VIEW) {
+        return false;
+      }
+    }
+
+    return Boolean(
+      embeddable.getRoot() &&
+        embeddable.getRoot().isContainer &&
+        embeddable.getRoot().type === DASHBOARD_CONTAINER_TYPE
+    );
+  }
+
+  public async execute({ embeddable }: ClonePanelActionContext) {
+    if (!embeddable.getRoot() || !embeddable.getRoot().isContainer) {
+      throw new IncompatibleActionError();
+    }
+
+    const dashboard = embeddable.getRoot() as DashboardContainer;
+    const panelToClone = dashboard.getInput().panels[embeddable.id] as DashboardPanelState;
+    if (!panelToClone) {
+      throw new PanelNotFoundError();
+    }
+
+    dashboard.showPlaceholderUntil(
+      this.cloneEmbeddable(panelToClone, embeddable.type),
+      placePanelBeside,
+      {
+        width: panelToClone.gridData.w,
+        height: panelToClone.gridData.h,
+        currentPanels: dashboard.getInput().panels,
+        placeBesideId: panelToClone.explicitInput.id,
+      } as IPanelPlacementBesideArgs
+    );
+  }
+
+  private async getUniqueTitle(rawTitle: string, embeddableType: string): Promise<string> {
+    const clonedTag = i18n.translate('dashboard.panel.title.clonedTag', {
+      defaultMessage: 'copy',
+    });
+    const cloneRegex = new RegExp(`\\(${clonedTag}\\)`, 'g');
+    const cloneNumberRegex = new RegExp(`\\(${clonedTag} [0-9]+\\)`, 'g');
```

Comment:
What exactly is the purpose of these regexes?

Reply:
They are used to remove any cloning artefacts from the name of the savedObject, so that you don't end up with a title like "Hello World (copy) (copy 1) (copy 2)"